pkg/dataobj/metastore/updater.go

```diff
@@ -168,39 +204,99 @@ func (m *Updater) Update(ctx context.Context, dataobjPath string, minTimestamp,
 	defer streamsReader.Close()
 
+	var indexPointersReader indexpointers.RowReader
+	defer indexPointersReader.Close()
+
 	// Read streams from existing metastore object and write them to the builder for the new object
 	buf := make([]streams.Stream, 100)
+	pbuf := make([]indexpointers.IndexPointer, 100)
 
 	for _, section := range object.Sections() {
-		if !streams.CheckSection(section) {
+		if !streams.CheckSection(section) && !indexpointers.CheckSection(section) {
 			continue
 		}
-		sec, err := streams.Open(ctx, section)
-		if err != nil {
-			return errors.Wrap(err, "opening section")
-		}
 
-		streamsReader.Reset(sec)
-		for n, err := streamsReader.Read(ctx, buf); n > 0; n, err = streamsReader.Read(ctx, buf) {
-			if err != nil && err != io.EOF {
-				return errors.Wrap(err, "reading streams")
+		switch {
+		// Backwards compatibility with old metastore top-level objects.
+		case streams.CheckSection(section):
+			sec, err := streams.Open(ctx, section)
+			if err != nil {
+				return MetaStoreTopLevelObjectTypeV1, errors.Wrap(err, "opening section")
 			}
-			for _, stream := range buf[:n] {
-				err = m.metastoreBuilder.Append(logproto.Stream{
-					Labels:  stream.Labels.String(),
-					Entries: []logproto.Entry{{Line: ""}},
-				})
-				if err != nil {
-					return errors.Wrap(err, "appending streams")
+
+			streamsReader.Reset(sec)
+			for n, err := streamsReader.Read(ctx, buf); n > 0; n, err = streamsReader.Read(ctx, buf) {
+				if err != nil && err != io.EOF {
+					return MetaStoreTopLevelObjectTypeV1, errors.Wrap(err, "reading streams")
 				}
+				for _, stream := range buf[:n] {
+					err = m.metastoreBuilder.Append(logproto.Stream{
+						Labels:  stream.Labels.String(),
+						Entries: []logproto.Entry{{Line: ""}},
+					})
+					if err != nil {
+						return MetaStoreTopLevelObjectTypeV1, errors.Wrap(err, "appending streams")
+					}
+				}
+			}
+
+			return MetaStoreTopLevelObjectTypeV1, nil
+		// New standard approach for metastore top-level objects.
+		case indexpointers.CheckSection(section):
+			sec, err := indexpointers.Open(ctx, section)
+			if err != nil {
+				return MetaStoreTopLevelObjectTypeV2, errors.Wrap(err, "opening section")
 			}
+			indexPointersReader.Reset(sec)
+			for n, err := indexPointersReader.Read(ctx, pbuf); n > 0; n, err = indexPointersReader.Read(ctx, pbuf) {
+				if err != nil && err != io.EOF {
+					return MetaStoreTopLevelObjectTypeV2, errors.Wrap(err, "reading index pointers")
+				}
+				for _, indexPointer := range pbuf[:n] {
+					err = m.builder.AppendIndexPointer(indexPointer.Path, indexPointer.StartTs, indexPointer.EndTs)
+					if err != nil {
+						return MetaStoreTopLevelObjectTypeV2, errors.Wrap(err, "appending index pointers")
+					}
+				}
+			}
+
+			return MetaStoreTopLevelObjectTypeV2, nil
+		default:
+			return MetaStoreTopLevelObjectTypeV2, errors.New("unknown section type")
 		}
 	}
 
-	return nil
+	return MetaStoreTopLevelObjectTypeV2, nil
```

Comment:
I think the default should still be V1 for now. It could be overriden with a flag or with a custom image for testing until the read path is merged & ready to go. To avoid disruption to the query path, we shouldn't build objects we can't read yet